src/main/scala/rise/core/DSL/package.scala

```diff
@@ -0,0 +1,485 @@
+                                       ): NatFunction4Wrapper[A] =
+    NatFunction4Wrapper(f)
+
+  case class NatFunction5Wrapper[A](f: (Nat, Nat, Nat, Nat, Nat) => A)
+  implicit def toNatFunction5Wrapper[A](f: (Nat, Nat, Nat, Nat, Nat) => A
+                                       ): NatFunction5Wrapper[A] =
+    NatFunction5Wrapper(f)
+
+  case class DataTypeFunctionWrapper[A](f: DataType => A)
+  implicit def toDataTypeFunctionWrapper[A](f: DataType => A
+                                           ): DataTypeFunctionWrapper[A] =
+    DataTypeFunctionWrapper(f)
+
+  case class NatToDataFunctionWrapper[A](f: NatToData => A)
+  implicit def toNatToDataFunctionWrapper[A](f: NatToData => A
+                                            ): NatToDataFunctionWrapper[A] =
+    NatToDataFunctionWrapper(f)
+
+  case class NatToNatFunctionWrapper[A](f: NatToNat => A)
+  implicit def toNatToNatFunctionWrapper[A](f: NatToNat => A
+                                           ): NatToNatFunctionWrapper[A] =
+    NatToNatFunctionWrapper(f)
+
+  case class AddressSpaceFunctionWrapper[A](f: AddressSpace => A)
+  implicit def toAddressSpaceFunctionWrapper[A](f: AddressSpace => A
+                                               ): AddressSpaceFunctionWrapper[A] =
+    AddressSpaceFunctionWrapper(f)
+
+  case class NatCollectionFunctionWrapper[A](f: NatCollectionIdentifier => A)
+  implicit def toNatCollectionFunctionWrapper[A](f: NatCollectionIdentifier => A
+                                                ): NatCollectionFunctionWrapper[A] =
+    NatCollectionFunctionWrapper(f)
+
+  implicit def wrapInNatExpr(n: Nat): ToBeTyped[Literal] = literal(NatData(n))
+
+  def l(i: Int): ToBeTyped[Literal] = literal(IntData(i))
+  def l(f: Float): ToBeTyped[Literal] = literal(FloatData(f))
+  def l(d: Double): ToBeTyped[Literal] = literal(DoubleData(d))
+  def l(n: Nat): ToBeTyped[Literal] = literal(NatData(n))
+  def lidx(i: Nat, n: Nat): ToBeTyped[Literal] = literal(IndexData(i, n))
+  def lvec(v: Seq[ScalarData]): ToBeTyped[Literal] = literal(VectorData(v))
+  def larr(a: Seq[Data]): ToBeTyped[Literal] = literal(ArrayData(a))
+
+  object foreignFun {
+    def apply(name: String, t: Type): ToBeTyped[ForeignFunction] = {
+      toBeTyped(ForeignFunction(ForeignFunction.Decl(name, None))(t))
+    }
+
+    def apply(
+               name: String,
+               params: Seq[String],
+               body: String,
+               t: Type
+             ): ToBeTyped[ForeignFunction] = {
+      toBeTyped(
+        ForeignFunction(
+          ForeignFunction.Decl(name, Some(ForeignFunction.Def(params, body)))
+        )(t)
+      )
+    }
+  }
+
+  // interface between ToBeTyped and Expr
+
+  implicit def toExpr[T <: Expr](d: ToBeTyped[T]): Expr = d.toExpr
+
+  implicit def preserveType[T <: Expr](e: T): ToBeTyped[Opaque] =
+    ToBeTyped(Opaque(e, e.t))
+
+  def toBeTyped[T <: Expr](e: T): ToBeTyped[T] = ToBeTyped(e)
+
+  def topLevel(e: Expr): TopLevel = TopLevel(e)()
```

Comment:
I don't really understand what this is supposed to model.
I think we should improve on our naming of it.

Reply:
I think that the idea was to avoid re-typing common expressions over and over again (e.g. if you use slide2D multiple times), and to infer a polymorphic type instead a bit like `let` bindings in OCaml, @roger-uw can correct me if this is wrong. I do not remember how this is different from `Opaque` though, for example can we do:
```
let definition: Expr = (bigExpression : ToBeTyped[Expr]).toExpr
let expression: ToBeTyped[Expr] = definition >> definition >> definition // definition becomes Opaque and is only typed once?
```

Reply:
@Bastacyclop Yes, `TopLevel` is designed to avoid repeated type inferences. It is named as `TopLevel` because we usually write a definition which will be used in multiple places as a Scala `val` or `def`, and it can be considered as a top-level definition for RISE. During type inference, the type of an `Opaque` expression will stay unchanged, but the free type variables captured in a `TopLevel` expression will be instantiated (like what happens to let-bindings in HM-style type inference).

Reply:
opened issue #79